### PR TITLE
[MDEP-799] - improve dependency:tree to add optional JSON output of the results

### PR DESCRIPTION
**Pull request by kezhenxu94**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MDEP) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MDEP-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MDEP-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

(I'm Apache Member and I've signed ICLA)

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.shared.dependency.graph.DependencyNode;
+import org.apache.maven.shared.dependency.graph.traversal.DependencyNodeVisitor;
+
+import java.io.Writer;
+import java.util.List;
+import com.google.common.base.Joiner;
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:30]: Isn't there a joiner with StringUtils?

Comment (kezhenxu94) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java]: > Isn't there a joiner with StringUtils?

@michael-o updated with `StringUtils`

File: src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java [hunk 2]
```diff
@@ -416,6 +416,10 @@ else if ( "dot".equals( outputType ) )
         {
             return new DOTDependencyNodeVisitor( writer );
         }
+        else if ( "json".equals( outputType ) )
+        {
+            return new JSONDependencyNodeVisitor( writer );
+        }
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/dependency/tree/TreeMojo.java:422]: can be replaced by `switch` statement

File: src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java [hunk 3]
```diff
@@ -141,9 +141,32 @@ public void _testTreeTGFSerializing()
         assertTrue( findString( contents, "testGroupId:release:jar:1.0:compile" ) );
     }
 
+    /**
+     * Test the JSON format serialization
+     *
+     * @throws Exception in case of an error.
+     */
+    public void _testTreeJSONSerializing()
```
Comment (slawekjaranowski) [inline src/test/java/org/apache/maven/plugins/dependency/tree/TestTreeMojo.java:149]: Please write dedicated test for new class `JSONDependencyNodeVisitor`

**Review by slawekjaranowski — CHANGES_REQUESTED** [review]
- why json is created manually? Maybe use FasterXML
- missing documentation, I propose separate page with output example and description

**Comment by kezhenxu94** [comment]
> * why json is created manually? Maybe use FasterXML

Is it OK to add some dependencies? I thought you might want to keep the dependencies minimal.

> * missing documentation, I propose separate page with output example and description

OK, will add doc

**Comment by slawekjaranowski** [comment]
> > * why json is created manually? Maybe use FasterXML
> 
> Is it OK to add some dependencies? I thought you might want to keep the dependencies minimal.
> 

It is ok if meets: https://maven.apache.org/developers/dependency-policies.html

> > * missing documentation, I propose separate page with output example and description
> 
> OK, will add doc

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ */
+public class JSONDependencyNodeVisitor
+    extends AbstractSerializingVisitor
+    implements DependencyNodeVisitor
+{
+
+    /**
+     * Constructor.
+     *
+     * @param writer the writer to write to.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:46]: nit: no period

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        return true;
+    }
+
+    private void writeNode( int indent, DependencyNode node, boolean root )
+    {
+        Artifact artifact = node.getArtifact();
+
+        writer.println( indentations( indent ) + "{" );
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:80]: We shouldn't be using a PrintWriter here. They swallow IOExceptions. Even if the superclass already does this, let's not make the same mistake again.

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        indent++;
+        String groupId = indentations( indent ) + "\"groupId\": \"" + artifact.getGroupId() + "\"";
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:82]: I think manually constructing the JSON is preferable to adding another library. (Some JSON libraries due have security issues.)

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        String artifactId = indentations( indent ) + "\"artifactId\": \"" + artifact.getArtifactId() + "\"";
+        String version = indentations( indent ) + "\"version\": \"" + artifact.getVersion() + "\"";
+        String type = indentations( indent ) + "\"type\": \"" + artifact.getType() + "\"";
+        String scope = indentations( indent ) + "\"scope\": \"" + artifact.getScope() + "\"";
+        String[] elements = root ? new String[] { groupId, artifactId, version, type }
+                                 : new String[] { groupId, artifactId, version, type, scope };
+
+        writer.print( StringUtils.join( "," + System.lineSeparator(), elements ) );
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:90]: Please don't system dependent output like System.lineSeparator(). Specify the character being output.

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java (hunk 7 above)
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:90]: Why, a `PrintWriter` will do the same?!

Comment (elharo) [reply src/main/java/org/apache/maven/plugins/dependency/tree/JSONDependencyNodeVisitor.java:90]: Yes, and that's another reason we shouldn't be using PrintWriter

**Comment by monperrus** [comment]
super cool feature, hope we can get it merged!

**Comment by MartinWitt** [comment]
Hey, as this PR seems stale, and I want this feature, I would implement this in a new PR. Do you prefer to use a JSON library, which one do you like, or shall I write a JSON printer myself?

**Comment by elharo** [comment]
I strongly prefer no extra dependencies for this. JSON libraries in particular are a world of security bugs, unmaintained code, violations of Java conventions, and overall poor design. For JSON output, writing strings is straight-forward. A library shouldn't be necessary to implement this.

Possibly tests (and only tests) might want to parse the JSON, and for that a library would be helpful. Do not use Jackson. javax.json or perhaps GSON might be OK.

**Comment by elharo** [comment]
> > I strongly prefer no extra dependencies for this. JSON libraries in particular are a world of security bugs, unmaintained code, violations of Java conventions, and overall poor design. For JSON output, writing strings is straight-forward. A library shouldn't be necessary to implement this.
> > Possibly tests (and only tests) might want to parse the JSON, and for that a library would be helpful. Do not use Jackson. javax.json or perhaps GSON might be OK.
> 
> GSON is superseded already.

By what?

**Comment by cstamas** [comment]
What makes you think that "gson is superseded"?

**Comment by michael-o** [comment]
> What makes you think that "gson is superseded"?

There is a while post about it.

**Comment by cstamas** [comment]
Where?

**Comment by michael-o** [comment]
> Where?

Searching, it was from the original authors, maybe even Kevin Bourillion.

**Comment by cstamas** [comment]
AFAIK, GSON was "just" moved out from Google umbrella, and is in maintenance mode, as it is feature complete (or something, dunno, guessing), but the project is quite lively and have regular releases and fixes.

**Comment by elharo** [comment]
While I was at Google, I did successfully lobby to get GSON adopted and officially supported by the Java core libraries team. That was a couple of years ago. Things might have changed since, but anything you heard about the project pre-2021 is certainly out of date.

**Comment by elharo** [comment]
feature complete under another PR

**Pull request closed without merging**